### ISSUE TITLE
VM service: resolve AsyncError for Isolate.error

### ISSUE THREAD

**Issue by nex3** [dart-lang/sdk#24163]
Currently, if an asynchronous error is bubbled up to the top level of an Isolate, the VM service's value for `Isolate.error.exception` contains an `AsyncError` object and its value for `Isolate.error.stackTrace` contains a few lines of internal code rather than the original exception's stack trace. It would be more useful if the protocol would automatically unwrap the `AsyncError` and present its exception as `Isolate.error.exception` and its stack trace as `Isolate.error.stackTrace`. This would also make `main() => throw "error";` and `main() async => throw "error";` end up in the same place from a client's perspective.

**Comment by johnmccutchan** [comment]
@rmacnak-google

**Comment by lrhn** [comment]
If the async code could report the uncaught error directly, by calling some "reportUncaughtError" VM function, then there wouldn't be any need for wrapping at all.

That's complicated by it still needing to tear down the isolate if errors are fatal, but at worst, it can wait until the next time code returns to the event handler.

**Comment by turnidge** [comment]
After looking at this, I think it is lossy for us to be in the business of unwrapping this object for the client.  I'd like to push back on this for now.

**Comment by turnidge** [comment]
If this ends up being important to our clients in the future it is possible to fix it in non-breaking ways.  For example, we could add optional fields somewhere for a wrappedException and a wrappedStacktrace.

**Comment by nex3** [comment]
> After looking at this, I think it is lossy for us to be in the business of unwrapping this object for the client.

Lossy how? From the perspective of a user, the entire process of the core libraries passing an asynchronous error to the VM is an implementation detail, so unwrapping it should be transparent—there's no difference between `dart:async` unwrapping it as @lrhn suggests or the service protocol doing the unwrapping.

**Comment by rmacnak-google** [comment]
Those programs produce different exceptions and the service protocol should faithfully report that. The protocol is very much in the business of showing the implementation of Dart programs - activation records, classes, functions, etc from core libraries are not hidden, nor should they be.

**Comment by nex3** [comment]
They only produce different exceptions inasmuch as the VM and the service protocol choose to treat them as such. The original `throw` statement certainly didn't throw an `AsyncError`; the internals chose to wrap it in the first place, and they could choose to unwrap them as well.

> If this ends up being important to our clients in the future it is possible to fix it in non-breaking ways. For example, we could add optional fields somewhere for a wrappedException and a wrappedStacktrace.

I'm a client, and it's important for me. It should be easy and uniform to access the exception and stack trace that the user actually threw, not some abstraction that the internal infrastructure uses to pass them around.

**Comment by lrhn** [comment]
Yes, this is very much the internal async loop at fault. When we capture an uncaught error inside event propagation, it has already been "reflected" into values (e.g., the error completion of a Future).
We can't rethrow this in such a way that the stack trace is reified. 
What we need in order to do that is a VM function that either allows us to throw an error *and* a stacktrace as if it had just been thrown, or that allows us to inject an error/stacktrace pair as if it had been caught by the top-level event loop. I would probably prefer the former (and maybe that's a functionality that other framework designers might want to be publicly available).

Since we have neither, we just fall back on throwing an Error object (AsyncError) which has the same toString as the orignal error and a stackTrace getter returning the original stackTrace. That AsyncError is throw from a fresh microtask, so its actual stack trace is pretty short.

In other words: Give us the functionality of throwing with a dynamically supplied stack trace object, and I'll fix this problem with a one-line change. :)

**Comment by lrhn** [comment]
Might actually be much simpler to fix than I had anticipated: https://codereview.chromium.org/1383983002/

**Comment by nex3** [comment]
@lrhn That's awesome!

For what it's worth, though, throwing with an existing stack would be very useful for a fair amount of code I've written.

**Comment by nex3** [comment]
I'm going to close this out, since da8baacae46e5f1f00d90ada1fa85dc0a9e0286c makes it moot.
